src/containers/barray.mli

```diff
@@ -0,0 +1,213 @@
+
+(* This file is free software, part of containers. See file "license" for more details. *)
+
+type 'a t = 'a array
+
+val empty : 'a t
+
+val isEmpty: 'a t -> bool
+
+external make : int -> 'a -> 'a t = "caml_make_vect"
+(** [Barray.make n x] returns a fresh t of length [n],
+    initialized with [x].
+    All the elements of this new t are initially
+    physically equal to [x] (in the sense of the [==] predicate).
+    Consequently, if [x] is mutable, it is shared among all elements
+    of the t, and modifying [x] through one of the t entries
+    will modify all other entries at the same time.
+    Raise [Invalid_argument] if [n < 0] or [n > Sys.max_array_length].
+    If the value of [x] is a floating-point number, then the maximum
+    size is only [Sys.max_array_length / 2].*)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [Barray.init n f] returns a fresh t of length [n],
+    with element number [i] initialized to the result of [f i].
+    In other terms, [Array.init n f] tabulates the results of [f]
+    applied to the integers [0] to [n-1].
+    Raise [Invalid_argument] if [n < 0] or [n > Sys.max_array_length].
+    If the return type of [f] is [float], then the maximum
+    size is only [Sys.max_array_length / 2].*)
+
+val length: 'a t -> int
+
+val get : int -> 'a t -> 'a option
+
+val set : int -> 'a -> 'a t -> bool
+
+val getOrRaise : int -> 'a t -> 'a
+
+val setOrRaise : int -> 'a -> 'a t -> unit
+
+val equals : 'a Equality.t -> 'a t -> 'a t -> bool
+
+val compare : ('a -> 'b -> Ordering.t) -> 'a t -> 'b t -> Ordering.t
+
+val append : 'a t -> 'a t -> 'a t
+(** [Barray.append v1 v2] returns a fresh t containing the
+    concatenation of the arrays [v1] and [v2]. *)
+
+val concat : 'a t list -> 'a t
+(** Same as [Barray.append], but concatenates a list of arrays. *)
+
+val slice : int -> int -> 'a t -> 'a t option
```

Comment:
I really dislike having two apis, one labelled one not. Smells of lack of decisiveness on the behalf of the API writer and is harder to learn.

I generally used name labels when they are needed to clarify the intent of the API signature. In this case, there is no way to know what the the various integer arguments do without labels.